velox/expression/fuzzer/tests/DecimalArgGeneratorTest.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include "velox/expression/SignatureBinder.h"
+#include "velox/expression/fuzzer/DecimalArgGeneratorBase.h"
+
+namespace facebook::velox::fuzzer::test {
+
+class DecimalArgGeneratorTest : public testing::Test {
+ protected:
+  class UnaryArgGenerator : public DecimalArgGeneratorBase {
+   public:
+    UnaryArgGenerator() {
+      initialize(1);
+    }
+
+   protected:
+    std::optional<std::pair<int, int>> toReturnType(int p, int s, ...)
+        override {
+      auto precision = std::min(38, p + s + 1);
+      auto scale = std::min(s + 1, 18);
+      return {{precision, scale}};
+    }
+  };
+
+  class BinaryArgGenerator : public DecimalArgGeneratorBase {
+   public:
+    BinaryArgGenerator() {
+      initialize(2);
+    }
+
+   protected:
+    std::optional<std::pair<int, int>> toReturnType(int p, int s, ...)
+        override {
+      va_list args;
```

Comment:
nit: I wonder if it would be more convenient to define 2 separate toReturnType methods and have the derived classes implement one of these. Otherwise, all 2+ arg derived classes would have to use va_list + va_start, which adds a few lines of boiler plate code.

```
virtual std::optional<std::pair<int, int>> toReturnType(int p, int s) {
   VELOX_UNREACHABLE();
}

virtual std::optional<std::pair<int, int>> toReturnType(int p1, int s1, int p2, int s2) {
   VELOX_UNREACHABLE();
}
```